contrib/src/main/scala/akka/stream/contrib/Valve.scala

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.stream.contrib
+
+import akka.stream.stage.{ GraphStageLogic, GraphStageWithMaterializedValue, InHandler, OutHandler }
+import akka.stream._
+import akka.stream.contrib.SwitchMode.{ Close, Open }
+
+import scala.concurrent.{ Future, Promise }
+
+/**
+ * Pause/ Resume a Flow
+ */
+sealed trait ValveSwitch {
+  /**
+   * Change the state of the valve
+   *
+   * @param mode expected mode to switch on
+   * @return A future that completes with true if the mode did change and false if it already was in the requested mode
+   */
+  def flip(mode: SwitchMode): Future[Boolean]
+}
+
+object Valve {
+  /**
+   * Factory for [[Valve]] instances.
+   *
+   */
+  def apply[A](): Valve[A] = Valve[A](SwitchMode.Open)
+
+  /**
+   * Java API: Factory for [[Valve]] instances.
+   *
+   */
+  def create[A](): Valve[A] = Valve[A](SwitchMode.Open)
+
+  /**
+   * Factory for [[Valve]] instances.
+   *
+   */
+  def apply[A](mode: SwitchMode): Valve[A] = new Valve[A](mode)
+
+  /**
+   * Java API: Factory for [[Valve]] instances.
+   *
+   */
+  def create[A](mode: SwitchMode): Valve[A] = Valve[A](mode)
+
+}
+
+/**
+ * Materializes into a [[ValveSwitch]] which provides a the method flip that stops or restarts the flow of elements passing through the stage. As long as the valve is closed it will backpressure.
+ *
+ * Note that closing the valve could result in one element being buffered inside the stage, and if the stream completes or fails while being closed, that element may be lost.
+ *
+ * @param mode state of the valve at the startup of the flow (by default Open)
+ */
+final class Valve[A](mode: SwitchMode) extends GraphStageWithMaterializedValue[FlowShape[A, A], ValveSwitch] {
+
+  val in: Inlet[A] = Inlet[A]("valve.in")
+
+  val out: Outlet[A] = Outlet[A]("valve.out")
+
+  override val shape = FlowShape(in, out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, ValveSwitch) = {
+    val logic = new ValveGraphStageLogic(shape, mode)
+    (logic, logic.switch)
+  }
+
+  private class ValveGraphStageLogic(shape: Shape, var mode: SwitchMode) extends GraphStageLogic(shape) with InHandler with OutHandler {
+    val switch = new ValveSwitch {
+
+      val callback = getAsyncCallback[(SwitchMode, Promise[Boolean])] {
```

Comment:
I was a bit surprised why an asyncCallback is needed instead of a dumb AtomicBoolean but I got it now: You might need to kick out the stage into execution when it has an element buffered. I guess this is the simplest thing that can work, but I am a bit unhappy with the Future[Boolean] return value.